[16.0][ADD] shopinvader_anonymous_partner, shopinvader_fastapi_auth_jwt

sbidoul opened the pull request:
See [USAGE.rst](https://github.com/shopinvader/odoo-shopinvader/blob/16.0-shopinvader_guest_partner/shopinvader_anonymous_partner/readme/USAGE.rst).

Towards #1368 

depends on
- [x] https://github.com/OCA/server-auth/pull/520
- [x] https://github.com/OCA/rest-framework/pull/347

sbidoul committed:
[ADD] shopinvader_anonymous_partner

sbidoul commented:
Should we use sudo to create anonymous partners (probably)?

Should we use sudo to get anonymous partners?

@lmignon any thought about that?

lmignon commented:
> Should we use sudo to create anonymous partners (probably)?

yes
> 
> Should we use sudo to get anonymous partners?
> 

Yes. In fact into the depends methods
* the `authenticated_partner_impl` return a partner into a sudo env
* the `authenticated_partner_env` create an env with the partner.id de `authenticated_partner_impl` dans le contexte
* the `authenticated_partner` return the `partner` from a browse into the  `authenticated_partner_env` of the `partner.id` from `authenticated_partner_impl`  






> @lmignon any thought about that?

sbidoul commented:
Ok, thanks. I changed to use sudo in create and get and added a warning note in the README.

This PR is now be ready to review. But we should probably wait to see how it works with the cart service before merging.

lmignon commented:
We just discovered with @marielejeune that the implementation of the `authenticated_partner_env` is not secure since it doesn't enforce that the partner env is initialized from the odoo_env.... I'll fix it.

lmignon commented on shopinvader_anonymous_partner/readme/ROADMAP.rst:
```diff
@@ -0,0 +1 @@
+- some cron to remove old anonymous partners
```
IMO we should also provide a way to filter out by default anonymous partners into the contact's views. And a pre defined filter to list all anonymous partners....

sbidoul replied:
There was also the idea of making anonymous partners inactive.

sebastienbeau replied:
anonymous partner should be inactive by default, we only active them when we fill real information.

sbidoul replied:
> anonymous partner should be inactive by default, we only active them when we fill real information.

@sebastienbeau ok, I'll update this PR to do that.

Regarding activation, I'm now thinking the normal flow should not activate an anonymous partner, but rather transfer its cart (or other objects) to another partner. I see two cases:

1. Anonymous partner authenticates as an already known customer: we transfer the cart to the known customer and drop the anonymous partner.
2. Anonymous partner signs up: in that case create a new partner with sign-up information, and proceed as in case 1.

sebastienbeau replied:
You are as fast as V16 "blink of an eye" to answer to my comment ;).

Transfer can be a solution indeed (so we have the same flow as login with an existing one). Maybe we can "drop" the anonymous partner after the transfer.

hparfr replied:
we implemented transfert here https://github.com/shopinvader/odoo-shopinvader/pull/1251

For the deletion of anonymous partner it can be done with a clean up cron or something.

hparfr replied:
> anonymous partner should be inactive by default, we only active them when we fill real information.

why ?
A user can set a lang, an adress, etc... it's real info btw.

sbidoul replied:
It makes sense to me to have anonymous partners inactive to avoid polluting the contacts list.
I suspect users will always reach the anonymous partner forms by navigating from the carts.

sbidoul committed:
[ADD] shopinvader_fastapi_auth_jwt

sbidoul commented:
@lmignon @marielejeune I added here the `shopinvader_fastapi_auth_jwt` module which provides two dependencies:

- `auth_jwt_authenticated_or_anonymous_partner_autocreate`: to be used the cart `/sync` route 
- `auth_jwt_authenticated_or_anonymous_partner`: to be used everywhere else in shopinvader

To avoid creating anonymous partners for visitors who don't add anything to their cart, the front should avoid syncing the cart when it has no pending transaction (I think the cart lib does that already, to be checked).

sbidoul committed:
[WIP] add unmerged depency to test requirements

sbidoul committed:
Add httpx to test-requirements

Require for FastAPI TestClient

sbidoul commented:
Ok, this should be ready.

To understand how all this fits together, the easiest is to read the [tests](https://github.com/shopinvader/odoo-shopinvader/blob/16.0-shopinvader_guest_partner/shopinvader_fastapi_auth_jwt/tests/test_auth_jwt_or_anonymous.py).

sbidoul committed:
[IMP] shopinvader_anonymous_partner: create inactive

sebastienbeau commented:
Really good work, thanks for it !

Just a minor question. In the module cart sync we do not depend on jwt module, so we should have somewhere a kind of abstraction

    auth_authenticated_or_anonymous_partner_autocreate: to be used the cart /sync route
    auth_authenticated_or_anonymous_partner: to be used everywhere else in shopinvader


Then depending of the auth implementation it will use jwt or something else

sbidoul commented:
> Just a minor question. In the module cart sync we do not depend on jwt module, so we should have somewhere a kind of abstraction

Yes there is such an abstraction. In general, shopinvader routes will depend on `authenticated_partner`, which has an overridable implementation (`authenticated_partner_impl`) to be defined at app construction time.

The idea is that the cart route will also depend on `authenticated_partner` but will receive a different override than the other services: `auth_authenticated_or_anonymous_partner_autocreate`. Examples coming soon in or next to https://github.com/shopinvader/odoo-shopinvader/pull/1369.

There will probably be a third variant for the `signup` route where we need a valid authentication, but accept a JWT payload that does not correspond to a known partner yet (so we can create a new partner from the info in the JWT payload, possibly populated by social auth).

But the whole authentication mechanism is configurable when assembling the routes in the app (as well as with some darker magic using `base_rest`).

sbidoul commented:
PR updated to create anonymous partners active=False and search them with active_test=False.

sbidoul committed:
[REF] shopinvader_anonymous_partner: rename methods

Add cookie in the method names, so people can build on the concept
without using cookies if needed.

sbidoul commented:
@lmignon @sebastienbeau should we set the language on the anonymous partner (from Accept-Language header) in this module ? Should we try to infer the country from the HTTP request ?

lmignon commented:
@sbidoul The language from  Accept-Language header is set into the env if it's installed into the odoo instance. You can therefore use-it for the partner.
IMO, for the country we should also infer the country from the ip like in odoo https://github.com/odoo/odoo/blob/16.0/addons/website_sale/controllers/main.py#L1296

lmignon reviewed (approved):
LGTM (Code review only)

hparfr commented on shopinvader_anonymous_partner/models/res_partner.py:
```diff
@@ -0,0 +1,87 @@
+
+    _inherit = "res.partner"
+
+    anonymous_token = fields.Char()
```
add some doc

hparfr replied:
index=true ?

sbidoul replied:
the `_sql_constraint` creates the index

hparfr replied:
by the way why an exclude instead of unique ? 
here you don't allow multiple rows of being `""` (empty)

sbidoul replied:
@hparfr I replaced with a UNIQUE constraint, and I added a help text on the token field.

hparfr commented on shopinvader_anonymous_partner/models/res_partner.py:
```diff
@@ -0,0 +1,87 @@
+    ]
+
+    @api.model
+    def _create_anonymous_partner__token(self):
```
what is the meaning of the double underscore ?

hparfr replied:
can we have the response in the args  in order to set some additionnal values like the (shopinvader) backend asked in the request ?

sbidoul replied:
> what is the meaning of the double underscore ?

That's because with a single underscore it would give the impression to create a token, while here it creates the partner and returns the token. The other creates the partner and sets the cookie. Well... I may be overthinking :)

sbidoul replied:
> can we have the response in the args in order to set some additionnal values like the (shopinvader) backend asked in the request ?

I don't understand your question. Can you elaborate / give an example?

hparfr replied:
Use case:  you have multiple web sites, you want to know on which website the user has registred. 

If you want to implement it; you can extract some info from the request and write it on the user.

sbidoul replied:
Ah, yes I see. But I think we should rather made that information available in the context, like we do for the `authenticated_partner_id`. Don't we have that already, some convention that gives access to the current `shopinvader.backend` ?

hparfr commented on shopinvader_anonymous_partner/models/res_partner.py:
```diff
@@ -0,0 +1,87 @@
+    def _get_anonymous_partner__cookie(self, cookies: Cookies):
+        token = cookies.get(COOKIE_NAME)
+        if not token:
+            return self.env["res.partner"].sudo().browse()
```
why sudo here ?

sbidoul replied:
Sudo because anonymous partners are typically created by public/unauthenticated services running with very low privieges.

sbidoul replied:
And the sudo in this empty browse is for symmetry with the record returned by the create below. So I can write in the documentation that all recordsets returned here are sudo'ed.

hparfr reviewed (commented):
minor remarks

sbidoul committed:
[IMP] anonymous_partner: add help

sbidoul committed:
[IMP] anonymous_partner: simplify unique constraint

hparfr reviewed (approved):
thanks

sbidoul commented:
/ocabot merge patch

sbidoul commented:
/ocabot merge patch

sbidoul commented:
/ocabot merge patch

sbidoul commented:
/ocabot merge patch
